mappers/gige-sdk/driver/gigeclient.go

```diff
@@ -0,0 +1,357 @@
+				gigEClient.deviceMeta[DeviceSN].deviceStatus = false
+				go gigEClient.ReConnectDevice(DeviceSN)
+			}
+			return err
+		}
+		return nil
+	}
+}
+
+func (gigEClient *GigEVisionDevice) Get(DeviceSN string) (results string, err error) {
+	if strings.EqualFold(gigEClient.deviceMeta[DeviceSN].FeatureName, "ImageTrigger") {
+		var imageBuffer *byte
+		var size int
+		var msg *C.char
+		defer C.free(unsafe.Pointer(msg))
+		signal := C.get_image(gigEClient.deviceMeta[DeviceSN].dev, C.CString(gigEClient.deviceMeta[DeviceSN].imageFormat), (**C.char)(unsafe.Pointer(&imageBuffer)), (*C.int)(unsafe.Pointer(&size)), &msg)
+		if signal != 0 {
+			errorMsg := fmt.Sprintf("Failed to get %s's images: %s", DeviceSN, (string)(C.GoString(msg)))
+			err = errors.New(errorMsg)
+			if signal >= 4 {
+				gigEClient.deviceMeta[DeviceSN].deviceStatus = false
+				go gigEClient.ReConnectDevice(DeviceSN)
+			}
+			return "", err
+		}
+		var buffer []byte
+		var bufferHdr = (*reflect.SliceHeader)(unsafe.Pointer(&buffer))
+		bufferHdr.Data = uintptr(unsafe.Pointer(imageBuffer))
+		bufferHdr.Len = size
+		bufferHdr.Cap = size
+		results = base64.StdEncoding.EncodeToString(buffer)
+		C.free_image((**C.char)(unsafe.Pointer(&imageBuffer)))
+	} else if strings.EqualFold(gigEClient.deviceMeta[DeviceSN].FeatureName, "ImageFormat") {
+		if gigEClient.deviceMeta[DeviceSN].imageFormat != "" {
+			results = gigEClient.deviceMeta[DeviceSN].imageFormat
+		} else {
+			errorMsg := fmt.Sprintf("Maybe init %s's image format failed, it only support format png or pnm.", DeviceSN)
+			err = errors.New(errorMsg)
+			return "", err
+		}
+	} else if strings.EqualFold(gigEClient.deviceMeta[DeviceSN].FeatureName, "ImageURL") {
+		if gigEClient.deviceMeta[DeviceSN].imageURL != "" {
+			results = gigEClient.deviceMeta[DeviceSN].imageURL
+		} else {
+			errorMsg := fmt.Sprintf("Maybe init %s's image format failed, it only support format png or pnm.", DeviceSN)
+			err = errors.New(errorMsg)
+			return "", err
+		}
+		gigEClient.PostImage(DeviceSN)
+		return results, nil
+	} else {
+		var msg *C.char
+		var value *C.char
+		defer C.free(unsafe.Pointer(msg))
+		defer C.free(unsafe.Pointer(value))
+		signal := C.get_value(gigEClient.deviceMeta[DeviceSN].dev, C.CString(gigEClient.deviceMeta[DeviceSN].FeatureName), &value, &msg)
+		if signal != 0 {
+			errorMsg := fmt.Sprintf("Get %s from device %s's failed: %s", gigEClient.deviceMeta[DeviceSN].FeatureName, DeviceSN, (string)(C.GoString(msg)))
+			err = errors.New(errorMsg)
+			if signal != 1 {
+				gigEClient.deviceMeta[DeviceSN].deviceStatus = false
+				go gigEClient.ReConnectDevice(DeviceSN)
+			}
+			return "", err
+		}
+		results = C.GoString(value)
+	}
+	return results, err
+}
+
+func (gigEClient *GigEVisionDevice) NewClient(DeviceSN string) (err error) {
+	var msg *C.char
+	var dev *C.uint
+	if gigEClient.deviceMeta == nil {
+		gigEClient.deviceMeta = make(map[string]*DeviceMeta)
+	}
+	if _, ok := gigEClient.deviceMeta[DeviceSN]; !ok {
```

Comment:
if ok {
    return nil
}
if xxx {
    return nil
}

<main body>
return nil

Reply:
Thank you for your valuable comments. The code has been modified. Please review it further.